File: pset6/sentimental-credit/credit.py
from cs50 import get_string


def valid_checksum(number):
    sum = 0
    tmp = number

    # Logic from previous implementation
    while True:
        if tmp <= 0:
            break

        at_twos = ((tmp % 100) // 10) * 2
        at_ones = (tmp % 100) % 10

        if (at_twos >= 10):
            sum += at_twos // 10
            sum += at_twos % 10
        else:
            sum += at_twos

        sum += at_ones
        tmp //= 100

    if sum % 10 == 0:
        return True

    print("INVALID")
    return False


# Prompt for input
number = get_string("Number: ")

# Calculate checksum
cardnum = int(number)

if valid_checksum(cardnum):
    length = len(number)
    first_two = int(number[:2])
    first = int(number[:1])

    if (first_two in [34, 37] and length == 15):
        print("AMEX")
    elif (first_two in range(51, 56) and length == 16):
        print("MASTERCARD")
    elif (first == 4 and length in [13, 16]):
        print("VISA")
    else:
        print("INVALID")

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>

int checkLength(long number);
bool checkSum(long number);
bool validLength(int length);

int main()
{
    // prompt for input
    // calculate checksum
    // check for card length and starting digits
    long cardNumber = get_long("Number: ");
    int length = checkLength(cardNumber);

    // printf("Checksum: %b\n", checkSum(cardNumber));

    if (checkSum(cardNumber) && validLength(length))
    {
        // printf("Length: %d\n", length);

        long temp = cardNumber;
        for (int i = 0; i < length - 2; i++)
        {
            temp /= 10;
        }

        if ((temp == 34 || temp == 37) && length == 15)
        {
            printf("AMEX\n");
        }
        else if ((temp > 50 && temp < 56) && length == 16)
        {
            printf("MASTERCARD\n");
        }
        else if ((temp > 39 && temp < 50) && (length == 13 || length == 16))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int checkLength(long number)
{
    int length = 0;
    long quotient = number;

    while (quotient != 0)
    {
        quotient /= 10;
        length++;
    }

    return length;
}

bool checkSum(long number)
{
    long tempNumber = number;
    int everyOtherSum = 0, untouched = 0, eosChecker, total;

    while (tempNumber != 0)
    {
        untouched += (tempNumber % 100) % 10;

        printf("untouched: %d\n", untouched);

        eosChecker = (tempNumber % 100 / 10) * 2;

        printf("eos: %d\n", eosChecker);
        if (eosChecker >= 10)
        {
            everyOtherSum += eosChecker % 10;
            everyOtherSum += eosChecker / 10;
        }
        else
        {
            everyOtherSum += eosChecker;
        }

        tempNumber /= 100;
    }

    total = everyOtherSum + untouched;

    printf("Total: %d\n", total);

    if (total % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool validLength(int length)
{
    if (length == 13 || length == 15 || length == 16)
    {
        return true;
    }
    return false;
}
